test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/CacheTagKeyTest.cs

```diff
@@ -0,0 +1,393 @@
+            Assert.Equal(expected, key);
+        }
+
+        [Theory]
+        [InlineData("Vary-By-Value")]
+        [InlineData("Vary  with spaces")]
+        [InlineData("  Vary  with more spaces   ")]
+        public void GenerateKey_UsesVaryByPropertyToGenerateKey(string varyBy)
+        {
+            // Arrange
+            var tagHelperContext = GetTagHelperContext();
+            var cacheTagHelper = new CacheTagHelper(Mock.Of<IMemoryCache>(), new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                VaryBy = varyBy
+            };
+            var expected = "CacheTagHelper||testid||VaryBy||" + varyBy;
+
+            // Act
+            var cacheTagKey = CacheTagKey.From(cacheTagHelper, tagHelperContext);
+            var key = cacheTagKey.GenerateKey();
+
+            // Assert
+            Assert.Equal(expected, key);
+        }
+
+        [Theory]
+        [InlineData("Cookie0", "CacheTagHelper||testid||VaryByCookie(Cookie0||Cookie0Value)")]
+        [InlineData("Cookie0,Cookie1",
+            "CacheTagHelper||testid||VaryByCookie(Cookie0||Cookie0Value||Cookie1||Cookie1Value)")]
+        [InlineData("Cookie0, Cookie1",
+            "CacheTagHelper||testid||VaryByCookie(Cookie0||Cookie0Value||Cookie1||Cookie1Value)")]
+        [InlineData("   Cookie0,   ,   Cookie1   ",
+            "CacheTagHelper||testid||VaryByCookie(Cookie0||Cookie0Value||Cookie1||Cookie1Value)")]
+        [InlineData(",Cookie0,,Cookie1,",
+            "CacheTagHelper||testid||VaryByCookie(Cookie0||Cookie0Value||Cookie1||Cookie1Value)")]
+        public void GenerateKey_UsesVaryByCookieName(string varyByCookie, string expected)
+        {
+            // Arrange
+            var tagHelperContext = GetTagHelperContext();
+            var cacheTagHelper = new CacheTagHelper(Mock.Of<IMemoryCache>(), new HtmlTestEncoder())
+            {
+                ViewContext = GetViewContext(),
+                VaryByCookie = varyByCookie
+            };
+            cacheTagHelper.ViewContext.HttpContext.Request.Headers["Cookie"] =
+                "Cookie0=Cookie0Value;Cookie1=Cookie1Value";
+
+            // Act
+            var cacheTagKey = CacheTagKey.From(cacheTagHelper, tagHelperContext);
+            var key = cacheTagKey.GenerateKey();
+
+            // Assert
+            Assert.Equal(expected, key);
+        }
+
+        [Theory]
+        [InlineData("Accept-Language", "CacheTagHelper||testid||VaryByHeader(Accept-Language||en-us;charset=utf8)")]
+        [InlineData("X-CustomHeader,Accept-Encoding, NotAvailable",
+            "CacheTagHelper||testid||VaryByHeader(X-CustomHeader||Header-Value||Accept-Encoding||utf8||NotAvailable||)")]
+        [InlineData("X-CustomHeader,  , Accept-Encoding, NotAvailable",
+            "CacheTagHelper||testid||VaryByHeader(X-CustomHeader||Header-Value||Accept-Encoding||utf8||NotAvailable||)")]
+        public void GenerateKey_UsesVaryByHeader(string varyByHeader, string expected)
```

Comment:
I don't know the exact answer. It's a more general question for the cache tag helper.